wasm/wat/parser.go

```diff
@@ -0,0 +1,248 @@
+	// reason, we cannot enforce source[0] == '(', and instead need to start the lexer to check the first token.
+	p.tokenParser = p.ensureLParen
+	line, col, err := lex(p.parse, p.source)
+	if err != nil {
+		return nil, &formatError{line, col, p.errorContext(), err}
+	}
+	return p.module, nil
+}
+
+// parse calls the delegate ModuleParser.tokenParser
+func (p *ModuleParser) parse(tok tokenType, tokenBytes []byte, line, col int) error {
+	return p.tokenParser(tok, tokenBytes, line, col)
+}
+
+func (p *ModuleParser) ensureLParen(tok tokenType, _ []byte, _, _ int) error {
+	if tok != tokenLParen {
+		return fmt.Errorf("expected '(', but found %s", tok)
+	}
+	p.tokenParser = p.startField
+	return nil
+}
+
+func (p *ModuleParser) startField(tok tokenType, tokenBytes []byte, _, _ int) (err error) {
+	if tok != tokenKeyword {
+		return fmt.Errorf("expected field, but found %s", tok)
+	}
+
+	// We expect p.currentField set according to a potentially nested "($fieldName".
+	// Each case must return a tokenParser that consumes the rest of the field up to the ')'.
+	// Note: each branch must handle any nesting concerns. Ex. "(module (import" nests further to "(func".
+	switch p.currentField {
+	case fieldInitial:
+		p.tokenParser, err = p.initialFieldHandler(tokenBytes)
+	case fieldModule:
+		p.tokenParser, err = p.moduleFieldHandler(tokenBytes)
+	case fieldModuleImport:
+		p.tokenParser, err = p.importFieldHandler(tokenBytes)
+	default:
+		return fmt.Errorf("unexpected current field %d", p.currentField)
+	}
+	return
+}
+
+// initialFieldHandler returns a tokenParser for the top-level fields in the WebAssembly source.
+func (p *ModuleParser) initialFieldHandler(fieldName []byte) (tokenParser, error) {
+	if string(fieldName) == "module" {
+		p.currentField = fieldModule
+		return p.parseModule, nil
+	} else {
+		return nil, fmt.Errorf("unexpected field: %s", string(fieldName))
+	}
+}
+
+func (p *ModuleParser) parseModule(tok tokenType, tokenBytes []byte, _, _ int) error {
+	switch tok {
+	case tokenID:
+		name := string(tokenBytes)
+		if p.module.name != "" {
+			return fmt.Errorf("redundant name: %s", name)
+		}
+		p.module.name = name
+	case tokenLParen:
+		p.tokenParser = p.startField // after this look for a field name
+		return nil
+	case tokenRParen: // end of module
+		p.currentField = fieldInitial
+	default:
+		return p.unexpectedToken(tok, tokenBytes)
+	}
+	return nil
+}
+
+func (p *ModuleParser) moduleFieldHandler(fieldName []byte) (tokenParser, error) {
+	switch string(fieldName) {
+	case "import":
+		p.currentField = fieldModuleImport
+		return p.parseImport, nil
+	case "start":
+		p.currentField = fieldModuleStart
+		return p.parseStart, nil
+	}
+	return nil, fmt.Errorf("unexpected field: %s", string(fieldName))
+}
+
+func (p *ModuleParser) parseImport(tok tokenType, tokenBytes []byte, _, _ int) error {
+	switch tok {
+	case tokenString: // Ex. "" or "foo" including quotes!
+		name := string(tokenBytes[1 : len(tokenBytes)-1]) // unquote
```

Comment:
maybe better add comment about we already ensure that len(tokenBytes) >= 2 in lexer?

Reply:
sure